### PR TITLE
[FLINK-23413][table] Port RelTimeIndicatorConverter from SCALA to JAVA.

### PR DESCRIPTION
**Pull request by beyond1920**
## What is the purpose of the change

The pr aims to port RelTimeIndicatorConverter from SCALA to JAVA.

## Brief change log

  - port RelTimeIndicatorConverter from SCALA to JAVA

## Verifying this change

  - Existed UT and IT

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by beyond1920** [comment]
@godfreyhe  @twalthr Please have a look at the pr when you have time.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java [hunk 1]
```diff
@@ -0,0 +1,812 @@
...
+
+    @Override
+    public RelNode visit(LogicalSort sort) {
+        return visitSimpleRel(sort);
+    }
+
+    @Override
+    public RelNode visit(LogicalMatch match) {
+        RelNode newInput = match.getInput().accept(this);
+        RexTimeIndicatorMaterializer materializer = new RexTimeIndicatorMaterializer(newInput);
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java:158]: the logic is changed: the previous behavior is using the unconverted input type, while the current behavior is using the converted input type.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java [hunk 2]
```diff
@@ -0,0 +1,812 @@
...
+                match.getSubsets(),
+                match.isAllRows(),
+                match.getPartitionKeys(),
+                match.getOrderKeys(),
+                newInterval);
+    }
+
+    @Override
+    public RelNode visit(RelNode node) {
+        if (node instanceof Collect) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java:220]: one suggestion is: we can extend RelShuttle and add `RelNode visit(Collect collect)`, `RelNode visit(LogicalWindowAggregate agg), etc..` to new shuffle, this class can extend from the new shuffle and avoid so many `if else`

Comment (beyond1920) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java:220]: If we introduce a new interface whichs extend `RelShuttle` like the interface `HiveRelShuttle`, We need override `accept` method in `RelNode` just like the `HiveFilter`, `HiveJoin`.
Do we really need to do this?
BTW, If we introduce a new interface which now matches `LogicalNode`, after we refactor `RelTimeIndicatorConverter` to work just before `physical_transform` phase, we need update the new  `RelShuttle` interface to match `FlinkLogicalNode` and update those `FlinkLogicalNode` to override `accept` method.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java (hunk 1 above)
Comment (beyond1920) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java:158]: Yes, the previous behavior is wrong. but it is a little hard when I try to find a case to cover this, so I skip the test case. I would try harder to find a case....

Comment (beyond1920) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java:158]: @godfreyhe, add a test `MatchRecognizeTest#testCascadeMatch` to cover this.

**Comment by beyond1920** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java (hunk 2 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/RelTimeIndicatorConverter.java:216]: make sense

**Review by godfreyhe — APPROVED** [review]
LGTM

**Comment by twalthr** [comment]
Thanks for the review @godfreyhe. I would also like to take a final look before merging. I would do this by tomorrow otherwise feel free to merge.

**Comment by godfreyhe** [comment]
@twalthr , I just check the Azure status (SUCCESS) and merge this pr, and DO NOT see your comments. So sorry about that. I will not close the JIRA until you confirm, feel free to comment.

**Comment by twalthr** [comment]
No worries @godfreyhe. I was also quite late to the party. Feel free to close the issue now. I quickly scanned the changed. I think the interesting piece will be the next subtask as this one is only a porting issue without touching tests.

**Comment by beyond1920** [comment]
@twalthr , the next subtask [pr](https://github.com/apache/flink/pull/16620) is ready now, please have a look, thanks a lot.

**Pull request closed without merging**